src/tree.rs

```diff
@@ -102,77 +135,148 @@ fn cond_swap(swap: bool, node: pallas::Base, sibling: pallas::Base) -> Pair {
 /// The layer with 2^n nodes is called "layer n":
 ///      - leaves are at layer MERKLE_DEPTH_ORCHARD = 32;
 ///      - the root is at layer 0.
-/// `l_star` is MERKLE_DEPTH_ORCHARD - layer - 1.
+/// `l` is MERKLE_DEPTH_ORCHARD - layer - 1.
 ///      - when hashing two leaves, we produce a node on the layer above the leaves, i.e.
-///        layer = 31, l_star = 0
-///      - when hashing to the final root, we produce the anchor with layer = 0, l_star = 31.
-fn hash_layer(l_star: usize, pair: Pair) -> pallas::Base {
+///        layer = 31, l = 0
+///      - when hashing to the final root, we produce the anchor with layer = 0, l = 31.
+fn hash_with_l(l: usize, pair: Pair) -> CtOption<pallas::Base> {
     // MerkleCRH Sinsemilla hash domain.
     let domain = HashDomain::new(MERKLE_CRH_PERSONALIZATION);
 
-    domain
-        .hash(
-            iter::empty()
-                .chain(i2lebsp_k(l_star).iter().copied())
-                .chain(
-                    pair.left
-                        .to_le_bits()
-                        .iter()
-                        .by_val()
-                        .take(L_ORCHARD_MERKLE),
-                )
-                .chain(
-                    pair.right
-                        .to_le_bits()
-                        .iter()
-                        .by_val()
-                        .take(L_ORCHARD_MERKLE),
-                ),
+    domain.hash(
+        iter::empty()
+            .chain(i2lebsp_k(l).iter().copied())
+            .chain(
+                pair.left
+                    .to_le_bits()
+                    .iter()
+                    .by_val()
+                    .take(L_ORCHARD_MERKLE),
+            )
+            .chain(
+                pair.right
+                    .to_le_bits()
+                    .iter()
+                    .by_val()
+                    .take(L_ORCHARD_MERKLE),
+            ),
+    )
+}
+
+/// A newtype wrapper for leaves and internal nodes in the Orchard
+/// incremental note commitment tree.
+///
+/// This wraps a CtOption<pallas::Base> because Sinsemilla hashes
+/// can produce a bottom value which needs to be accounted for in
+/// the production of a Merkle root. Leaf nodes are always wrapped
+/// with the `Some` constructor.
```

Comment:
```suggestion
/// the production of a Merkle root. Leaf nodes are always "some".
```
It's not literally a `Some` constructor for `CtOption`.

Reply:
No, but it's the semantic equivalent.